pages/instructorFileBrowser/instructorFileBrowser.js

```diff
@@ -297,47 +297,59 @@ function browseDirectory(file_browser, callback) {
-                  } else if (type.mime === 'application/pdf') {
-                    file_browser.isPDF = true;
-                  }
-                } else {
-                  debug(`could not get file type`);
-                }
-              })(callback);
-            } else {
-              debug(`found a text file`);
-              file_browser.isText = true;
-              file_browser.contents = hljs.highlightAuto(contents.toString('utf8')).value;
-              callback(null);
+      async () => {
+        const isBinary = await isBinaryFile(file_browser.paths.workingPath);
+        file_browser.isBinary = isBinary;
+        if (isBinary) {
+          const type = await FileType.fromFile(file_browser.paths.workingPath);
+          if (type) {
+            if (type?.mime.startsWith('image')) {
+              file_browser.isImage = true;
+            } else if (type?.mime === 'application/pdf') {
+              file_browser.isPDF = true;
             }
-          },
-          (err) => {
-            if (ERR(err, callback)) return;
-            callback(null); // should never get here
           }
-        );
+        } else {
+          // This is probably a text file. If it's is larger that 1MB, don't
+          // attempt to read it; treat it like an opaque binary file.
+          const { size } = await fs.stat(file_browser.paths.workingPath);
+          if (size > 1 * 1024 * 1024) {
+            file_browser.isBinary = true;
+            return;
+          }
+
+          file_browser.isText = true;
+
+          const contents = await fs.readFile(file_browser.paths.workingPath);
+          const stringifiedContents = contents.toString('utf8');
+
+          // Try to guess the language from the file extension. This takes
+          // advantage of the fact that Highlight.js includes common file extensions
+          // as aliases for each supported language, and `getLanguage()` allows
+          // us to look up a language by its alias.
+          //
+          // If we don't get a match, we'll try to guess the language by running
+          // `highlightAuto()` on the first few thousand characters of the file.
+          //
+          // Note that we deliberately exclude `ml` and `ls` from the extensions
+          // that we try to guess from, as they're ambiguous (OCaml/Standard ML
+          // and LiveScript/Lasso, respectively). For more details, see
+          // https://highlightjs.readthedocs.io/en/latest/supported-languages.html
+          let language = null;
+          const extension = path.extname(file_browser.paths.workingPath).substring(1);
+          if (!['ml', 'ls'].includes(extension) && hljs.getLanguage(extension)) {
+            language = extension;
+          } else {
+            const result = hljs.highlightAuto(stringifiedContents.slice(0, 2000));
+            language = result.language;
+          }
+          file_browser.contents = hljs.highlight(stringifiedContents, { language }).value;
```

Comment:
This is causing an error for empty files, since in this case hljs sets the language to undefined, which is not accepted as input to highlight. Given that we're calling highlight on the whole string anyway, doesn't it make more sense to just call `highlightAuto` on the full string as well?

Reply:
That's exactly the original problem we're trying to avoid: the implementation for `highlightAuto` is roughly "pass the string through every single grammar we have, compute a relevance score for each one, and pick the best one". For large-ish files, that results in a ton of synchronous compute and actually makes the node process run out of memory since it's buffering the results for each of the 300+ grammars in memory. That's why we only run `highlightAuto` on the first 2000 bytes.

Great find with an `undefined` language though. I think the correct fix there is to just always fall back to `plaintext` if we can't auto-detect a language

Reply:
Fixed in #6270, commit 21ea4ab.